### DISCUSSION TITLE
[Markdown] Specify theme context for images (Beta)

### DISCUSSION THREAD

**Discussion by dipree** [community/community#16910] [Other Feature Feedback, Questions, & Ideas]
You can now specify the theme an image is displayed for in Markdown using the HTML `<picture>` element in combination with the `prefers-color-scheme` media feature.

**For example:**

```HTML
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://user-images.githubusercontent.com/25423296/163456776-7f95b81a-f1ed-45f7-b7ab-8fa810d529fa.png">
  <img alt="Shows an illustrated sun in light color mode and a moon with stars in dark color mode." src="https://user-images.githubusercontent.com/25423296/163456779-a8556205-d0a5-45e2-ac17-42d089e3c3f8.png">
</picture>
```

**Renders as:**
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://user-images.githubusercontent.com/25423296/163456776-7f95b81a-f1ed-45f7-b7ab-8fa810d529fa.png">
  <img alt="Shows an illustrated sun in light color mode and a moon with stars in dark color mode." src="https://user-images.githubusercontent.com/25423296/163456779-a8556205-d0a5-45e2-ac17-42d089e3c3f8.png">
</picture>

We released a first iteration of this functionality last year and received a lot of great feedback from you. We reviewed again the possibilities and concluded that using the HTML `<picture>` construct is the best option from multiple points of view. We came up with a solution that makes it work with the GitHub theming system which was the initial concern.

Thank you for all the great feedback so far and let us know if there's anything that isn't quite working as expected.

:octocat: 🐳

**Comment by bric3** [comment]
Hello I noticed that contextual paths are not rewritten with the new `picture` element

Before one could write this
```html
<img src=".github/images/colorful.svg?raw=true#gh-light-mode-only" width="100%" height="100">
<img src=".github/images/gray.svg?raw=true#gh-dark-mode-only" width="100%" height="100">
```

and paths would be prepended with the current branch context, eg:

```html
<img src="/owner/repo/raw/current-branch/.github/images/gray.svg?raw=true#gh-dark-mode-only" width="100%" height="100" style="max-width: 100%;">
```

However, the image paths are not prepended with `picture`. It is necessary to prepend manually for now.

```html
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="/owner/repo/raw/current-branch/.github/images/gray.svg?raw=true" width="100%" height="100">
  <img src="/owner/repo/raw/current-branch/.github/images/colorful.svg?raw=true" width="100%" height="100">
</picture>
```

**Reply by bric3** [reply]
@dipree any feedback ?

**Reply by dipree** [reply]
We are looking into fixing it!

**Reply by bric3** [reply]
Thanks @dipree

**Reply by bric3** [reply]
It seems to work in the browser, so that's good 👍
Not yet in the iOS app (1.67.0).

**Reply by bric3** [reply]
And it’s broken again, when seeing the `master` on an url like `https://github.com/owner/repo`

<img width="833" alt="image" src="https://user-images.githubusercontent.com/803621/171954790-239df01b-0040-467c-9aa5-dc7f952d08bd.png">

Of course, ending `/`, eg `https://github.com/owner/repo/`, makes no difference.

**Reply by Counselor8** [reply]
Tanks

**Reply by lumaxis** [reply]
@bric3 This should work now! Would you mind giving it another try to see if you can use themed images with relative links?

**Reply by Gargron** [reply]
Works now!

**Reply by bric3** [reply]
Confirmed, it works.

**Reply by BOTCAHX** [reply]
Feedback

**Reply by daikon100** [reply]
its very awesome

**Reply by flynn31** [reply]
Thanks

**Comment by vinkla** [comment]
It seems `<img src="..." width="100%" />` has stopped working since this update. The width is now ignored, is this the expected behaviour?

**Reply by lumaxis** [reply]
Hm, I cannot reproduce this. Using `width="100%"` still seems to work as expected:

<img alt=“Cartoon drawing of a sun” src="https://github.blog/wp-content/uploads/2022/04/Security-Product@2x.png" width="100%">

**Reply by vinkla** [reply]
You're right. It seems it only affects GIFs. For example:

<img src="https://media.giphy.com/media/xiSZGYjpYgzkipNS7x/giphy.gif" width="100%">

**Reply by lumaxis** [reply]
@vinkla Ah! That _could_ be related to a different feature that recently shipped: https://github.blog/changelog/2022-05-19-option-to-prevent-animated-images-from-playing-automatically/

If you could [open a new discussion](https://github.com/github/feedback/discussions/new?category=accessibility-feedback) with this feedback, that would be super helpful 🙏🏼

**Reply by dipree** [reply]
This was fixed.

**Reply by daikon100** [reply]
definately

**Comment by kellertuer** [comment]
This is a really nice feature, but it seems the loading does not always work? I added it to our Readme here

https://github.com/JuliaManifolds/Manifolds.jl/blob/kellertuer/dark-logo-readme/README.md

But I had to switch from `blob/` which worked in the previous version to `raw` such that the loading worked correctly; it seems the old readme (without the dark mode modifier) was able to rewrite that url?

**Comment by bubba-123** [comment]
<fycollector> You can now specify the theme an image is displayed for in Markdown using the HTML `<picture>` element in combination with the `prefers-color-scheme` media feature.
> 
> **For example:**
> 
> ```
> <picture>
>   <source media="(prefers-color-scheme: dark)" srcset="https://user-images.githubusercontent.com/25423296/163456776-7f95b81a-f1ed-45f7-b7ab-8fa810d529fa.png">
>   <img alt=“Shows an illustrated sun in light color mode and a moon with stars in dark color mode.” src="https://user-images.githubusercontent.com/25423296/163456779-a8556205-d0a5-45e2-ac17-42d089e3c3f8.png">
> </picture>
> ```
> 
> **Renders as:** ![Shows an illustrated sun in light color mode and a moon with stars in dark color mode.](https://user-images.githubusercontent.com/25423296/163456779-a8556205-d0a5-45e2-ac17-42d089e3c3f8.png)
> 
> We released a first iteration of this functionality last year and received a lot of great feedback from you. We reviewed again the possibilities and concluded that using the HTML `<picture>` construct is the best option from multiple points of view. We came up with a solution that makes it work with the GitHub theming system which was the initial concern.
> 
> Thank you for all the great feedback so far and let us know if there's anything that isn't quite working as expected.
> 
> :octocat: 🐳

I Rather teach one pray then men to teach &@bubba-123/

**Comment by bric3** [comment]
It seems the iOS app does not yet support this feature.

<img alt="github app screenshot" width="50%" src="https://user-images.githubusercontent.com/803621/185569718-cd897271-07ac-4578-9436-bb0b8af5b66b.png">

**Reply by Andrew6rant** [reply]
Are you sure? It is working for me (iOS 14.3, Github 1.68.0, iPhone10,6).

**Reply by bric3** [reply]
Yes. I see the image placeholder while the web version works.
Although it could be related to the bug mentioned in [this discussion](https://github.com/orgs/github-community/discussions/16910#discussioncomment-2787270) but on iOS.

**Reply by Counselor8** [reply]
yes suitable @dipree

**Reply by daikon100** [reply]
its awesome

**Reply by bric3** [reply]
Still not fixed in github mobile app, while the feature is GA

**Reply by jhildenbiddle** [reply]
I can confirm that relative links used within a `<picture>` element do not work property in the iOS app. I've created a support ticket and hope to see the issue addressed.

**Reply by bric3** [reply]
Any news on this ?

**Comment by mahozad** [comment]
Is it possible to add ability to use relative image links?

This is possible for "regular" markdown image syntax but doesn't seem to work for this new approach.

For example, if we are referencing an image in *assets* subdirectory of where the markdown file is located,
we can do this with regular markdown image syntax like this:

```markdown
![My image description](assets/logo.svg)
```

**Reply by bric3** [reply]
As far as I know this now works, see [this thread](https://github.com/orgs/github-community/discussions/16910#discussioncomment-2787270).

**Comment by daikon100** [comment]
sure,you can

**Comment by duongdam** [comment]
and what about svg render?

**Comment by daikon100** [comment]
i would also love to have that also

bad

**Comment by mahozad** [comment]
> **For example:**
> 
> ```
> <picture>
>   <source media="(prefers-color-scheme: dark)" srcset="https://user-images.githubusercontent.com/25423296/163456776-7f95b81a-f1ed-45f7-b7ab-8fa810d529fa.png">
>   <img alt=“Shows an illustrated sun in light color mode and a moon with stars in dark color mode.” src="https://user-images.githubusercontent.com/25423296/163456779-a8556205-d0a5-45e2-ac17-42d089e3c3f8.png">
> </picture>
> ```

@dipree
The example uses illegal curly quotation marks `“...”` instead of straight quotation marks `"..."` for `alt` attribute of the `img` element.

**Comment by TaronVardanyan** [comment]
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://user-images.githubusercontent.com/25423296/163456776-7f95b81a-f1ed-45f7-b7ab-8fa810d529fa.png">
  <img alt=“Shows an illustrated sun in light color mode and a moon with stars in dark color mode.” src="https://user-images.githubusercontent.com/25423296/163456779-a8556205-d0a5-45e2-ac17-42d089e3c3f8.png">
</picture>

**Reply by TaronVardanyan** [reply]
Cool stuff :)

**Comment by romilbhardwaj** [comment]
Sorry if this is a duplicate of [#1583](https://github.com/github/markup/issues/1583).

GitHub seems to set the `prefers-color-scheme` CSS property to `dark` when using the light default in Night theme. Using the example from [github docs](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#specifying-the-theme-an-image-is-shown-to):
```
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://user-images.githubusercontent.com/25423296/163456776-7f95b81a-f1ed-45f7-b7ab-8fa810d529fa.png">
  <source media="(prefers-color-scheme: light)" srcset="https://user-images.githubusercontent.com/25423296/163456779-a8556205-d0a5-45e2-ac17-42d089e3c3f8.png">
  <img alt="Shows an illustrated sun in light color mode and a moon with stars in dark color mode." src="https://user-images.githubusercontent.com/25423296/163456779-a8556205-d0a5-45e2-ac17-42d089e3c3f8.png">
</picture>
```
This renders as:
![image](https://user-images.githubusercontent.com/4416605/184545845-4b4745e5-8dc2-433d-94ee-e781b8d4a2a3.png)

The expected output should have been:
![image](https://user-images.githubusercontent.com/4416605/184545864-a397de21-767f-4f6f-8205-dfe432493acd.png)



# Replication
This happens only when you use the following settings:
1. System OS is in dark mode
2. In github settings -> appearance: Theme mode: `Sync with system`
3. Light default is selected for the Night theme

Here is the configuration:
![image](https://user-images.githubusercontent.com/4416605/184545614-fba118f8-1dce-4934-9ab2-cab13056bfcc.png)

When this configuration is used, the github page sets `prefers-color-scheme: dark` even though the page is rendered in light mode.
![image](https://user-images.githubusercontent.com/4416605/184546009-ac9e1242-abae-47ce-a936-2dd24cb549c1.png)

**Reply by simonprovost** [reply]
+1! So greatly explained and no answer yet? UP, please 🙏 3-ish years since this message, any workaround?

This is very frustrating! Happened in one of our PR and thought it was our fault, appears not to be actually!

@dipree

**Comment by bric3** [comment]
@dipree Good news it also works for Asciidoc documents if using the HTML passthrough

```asciidoc

++++
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://user-images.githubusercontent.com/25423296/163456776-7f95b81a-f1ed-45f7-b7ab-8fa810d529fa.png">
  <img alt="Shows an illustrated sun in light color mode and a moon with stars in dark color mode." src="https://user-images.githubusercontent.com/25423296/163456779-a8556205-d0a5-45e2-ac17-42d089e3c3f8.png">
</picture>
++++

```

This should also work with images contextual to a repository, say `src=".github/images/dark.png"`.
